### PR TITLE
Makes ListTargets() much much faster.

### LINKED ISSUE #19855: ListTargets is Incredibly Expensive
Not too hard to see this; just profile the server--it's outpacing nearly everything, up to, and including LINDA.

Source is likely the 150+ hostile mobs on lavaland.

Server CPU, at rest is 16-25, but is typically 40-50 because of this proc, alone (or rather it being called so much).

### PR DESCRIPTION
**Pull request by MrsTonedOne**
fixes #19855
profiles: (A is original, the rest are various variations I thought up)

```
/mob/living/simple_animal/hostile/proc/ListTargetsG    42.008      118.524      118.559       195594
/mob/living/simple_animal/hostile/proc/ListTargetsF    27.078       42.788       42.813       195267
/mob/living/simple_animal/hostile/proc/ListTargetsE    32.423       48.520       48.540       195401
/mob/living/simple_animal/hostile/proc/ListTargetsD    46.633      118.919      118.938       195455
/mob/living/simple_animal/hostile/proc/ListTargetsC    63.589       90.622       90.646       195630
/mob/living/simple_animal/hostile/proc/ListTargetsB    62.156       88.851       88.880       195159
/mob/living/simple_animal/hostile/proc/ListTargetsA   382.978      383.049      383.087       195578
/mob/living/simple_animal/hostile/proc/ListTargets      4.748      895.996      896.142      1368084
```

i went with E because F had a few edge cases that weren't handled correctly.

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
10% the cost....I'll take it.

**Comment by xxalpha** [comment]
:dash:

**Comment by RemieRichards** [comment]
typecaches save the day again~
Nah this is really good stuff :+1:

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 1]
```diff
@@ -83,26 +83,21 @@
...
 
-		var/hostile_machines = list()
-		for(var/mecha in mechas_list)
-			hostile_machines += mecha
-		for(var/obj/machinery/porta_turret/P in machines)
-			hostile_machines += P
-		for(var/M in hostile_machines)
-			if(get_dist(M, targets_from) <= vision_range && can_see(targets_from, M, vision_range))
-				. += M
+		var/static/hostile_machines = typecacheof(/obj/machinery/porta_turret, /obj/mecha)
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: <i>nag</i> this should be ```var/static/list/hostile_machines``` :P 
(or we could do what /vg/ did with their ```string``` ```int``` etc. defines and have ```#define typecache list```)

Comment (MrStonedOne) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: fug! I'll have to re test now!

Comment (MrStonedOne) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: Profile results were about the same.

Comment (MrStonedOne) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: as for the list thing, it's fine like this (I forgot and don't want to make another commit) the type is only used for checking . operations at compile time, since i don't do . operations on the list, it doesn't actually change anything

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: I know it's uses, I just like being pedantic in my definitions :)

**Comment by RemieRichards** [comment]
also @MrStonedOne, of course the results were about the same, you made the list static! the list() op only happens once.

**Comment by MrStonedOne** [comment]
Well ya, but I had to be sure. Its unscientific to assume.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 2]
```diff
@@ -83,26 +83,21 @@
 
 //////////////HOSTILE MOB TARGETTING AND AGGRESSION////////////
 
-
 /mob/living/simple_animal/hostile/proc/ListTargets()//Step 1, find out what we can see
 	. = list()
 	if(!search_objects)
 		var/list/Mobs = hearers(vision_range, targets_from) - src //Remove self, so we don't suicide
```
Comment (optimumtact) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: why is this using hearers?

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: probably some ancient rumour about ```hearers()``` being slightly more performant? something spooky like that probably.

Comment (optimumtact) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: oh it's a builtin

interesting, this means they would ignore things that byond thinks can't hear you, which implies blocked by opaque objects

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: yes but ```view()``` is the same (or rather, ```viewers()```)
it's probably ancient voodoo as I said.

Comment (optimumtact) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: why not use viewers or hearers for the typecachefilterlist call instead as well then?

would save a can_see check?

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: apparently, and this is obvious if we actually look at what the code is assuming here, ```viewers()``` and ```hearers()``` are mobs only, meaning they wouldn't find mecha or turrets.

Comment (optimumtact) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: oh yeah that makes sense, my derp.

Thanks

Comment (MrStonedOne) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: there is much more wrong with this proc then what oranges pointed out, i decided against changing anything function wise about the proc.

This could be changed, to a type cached view or hear() search, merged with the mech/turret stuff, and get rid of can_hear() to be much much faster (most likely, i never profiled it)

The only issue is that can_see()/can_hear() and view()/hear() calculates things differently, and that would mark a change in functionality, so i didn't include it here as i have no idea why it does things the way it does.
